### DISCUSSION TITLE
How to correctly use compile_commands.json for building with different targets

### DISCUSSION THREAD

**Discussion by mikaelspringer** [clangd/clangd#1504] [Q&A] [answered]
Our code base can be built using different targets, resulting in different build artifacts. In the code, this is done with a lot of ifdefs. This means parts of the code is not relevant when developing for a certain target, resulting in a lot of errors and warnings in my editor. I can generate different compile_commands.json for each target, but is this the "right" way of handling this? I'm using Emacs and lsp-mode. Changing compile_commands.json is not a problem. I do not however see that the indexer picks up this change fully, which means the project is in part incorrectly indexed. How can I solve this?

**Comment by HighCommander4** [comment, marked as answer]
Switching between `compile_commands.json` files for different targets is the main approach that comes to mind. You may need to restart clangd to get it to reindex files in the new configuration.

An idea that may be interesting to explore (but may come with performance implications) would be to index the project in multiple configurations and merge the index contents; some tools like [Searchfox](searchfox.org/) (not based on clangd) do this.

If you have entire files which are specific to a target (as opposed to shared files with `#ifdef` blocks), you could prepare a combined `compile_commands.json` file which includes entries for all such files (while, for shared files, you'd have to pick a particular target to use, and that would govern which `#ifdef` branches are active vs. inactive). You might run into a few issues like https://github.com/clangd/clangd/issues/1258 (if different targets define symbols with the same name), but the overall experience is likely to be improved.

**Reply by HighCommander4** [reply]
> If you have entire files which are specific to a target

I should add, if you have entire **directories** specific to a target (for example, something like `project/target1`, `project/target2`, etc.), then if you can arrange for the target-specific directories to have their own `compile_commands.json` files (with shared files being listed in `project/compile_commands.json`), then you don't need to manually combine the files, things should just work (i.e. clangd will use the target-specific `compile_commands.json` for files in those directories because it searches for the closest ancestor directory with a `compile_commands.json` file).

**Reply by mikaelspringer** [reply]
Thanks for the answer. I'll explore this further.
